pkg/controllers/nodeclaim/tagging/controller.go

```diff
@@ -0,0 +1,179 @@
+)
+
+const tagPeriod = 1 * time.Second
+
+type Controller struct {
+	kubeClient    client.Client
+	ec2           *ec2.EC2
+	defaultTagRgx *regexp.Regexp
+	timer         *time.Timer
+}
+
+func NewController(kubeClient client.Client, sess *session.Session) *Controller {
+	return &Controller{
+		kubeClient:    kubeClient,
+		ec2:           ec2.New(sess),
+		defaultTagRgx: regexp.MustCompile(corev1beta1.NodePoolLabelKey + "/.*"),
+		timer:         time.NewTimer(tagPeriod),
+	}
+}
+
+func (c *Controller) Name() string {
+	return "ec2-instance-tagger"
+}
+
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	nodeClaim := &corev1beta1.NodeClaim{}
+	if err := c.kubeClient.Get(ctx, req.NamespacedName, nodeClaim); err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Extract resource ID from node spec
+	resourceID := path.Base(nodeClaim.Status.ProviderID)
+	if resourceID == "." || resourceID == "/" {
+		return reconcile.Result{}, fmt.Errorf("failed to extract resource ID from providerID")
+	}
+
+	// Check if instance has already been tagged
+	if isTagged, err := c.isInstanceTagged(ctx, nodeClaim, resourceID); err != nil || isTagged {
+		return reconcile.Result{}, err
+	}
+
+	// Tag instance
+	err := c.tagInstance(ctx, nodeClaim, resourceID)
+	if err != nil {
+		return reconcile.Result{Requeue: true}, client.IgnoreNotFound(err)
+	}
+
+	logging.FromContext(ctx).
+		With("node", req.NamespacedName.Name).
+		With("instanceID", resourceID).
+		Debugf("updated ec2 instance Name tag")
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Builder(_ context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(
+		controllerruntime.
+			NewControllerManagedBy(m).
+			For(&corev1beta1.NodeClaim{}).
+			WithEventFilter(onlyTaggableInstances()),
+	)
+}
+
+func onlyTaggableInstances() predicate.Predicate {
+	return predicate.NewPredicateFuncs(func(object client.Object) bool {
+		nodeClaim := object.(*corev1beta1.NodeClaim)
+
+		// Instance has already been tagged
+		if val, ok := nodeClaim.Annotations[v1beta1.AnnotationInstanceTagged]; ok && val == "true" {
+			return false
+		}
+
+		// Node name is not yet known
+		if nodeClaim.Status.NodeName == "" {
+			return false
+		}
+
+		return true
+	})
+}
+
+func (c *Controller) isInstanceTagged(ctx context.Context, nodeClaim *corev1beta1.NodeClaim, resourceID string) (bool, error) {
+	tag, err := c.getTagValue(resourceID, "Name")
+	if err != nil {
+		return false, err
+	}
+
+	// Check if the instance has the default tag (karpenter.sh/provisioner-name/*). If so it has either:
+	//   a) Already been tagged by us but it is not reflected on the API server
+	//   b) Tagged by some other process. In this case we will respect that tag.
+	if !c.defaultTagRgx.MatchString(tag) {
+		nodeClaim.Annotations[v1beta1.AnnotationInstanceTagged] = "true"
+		err := c.kubeClient.Update(ctx, nodeClaim)
```

Comment:
We should consider using a `Patch` operation here since this call can fail due to a conflict with [optimistic locking](https://www.learnsteps.com/what-is-optimistic-concurrency-how-does-is-it-help-to-scale-the-kubernetes-cluster/). See the [NodeTemplate controller](https://github.com/aws/karpenter/blob/main/pkg/controllers/nodeclass/controller.go#L71) for a good example of how we can check if things are changed and then do a Patch operation if things updated.

Reply:
Since we are looking at updating this annotation in two different places in two different functions, we should maybe even look at pulling out the `Patch` operation from this function and putting it at the end of the `Reconcile` loop so that we are patching once at the end of the reconcile.